Upgrade to latest Prepack v0.2.38

joseluisq opened the pull request:
This PR Upgrade to latest Prepack `v0.2.38` according to exposed on #13 thread.
Feedback welcome.

Commit:
upgrade to latest prepack v0.2.38 (fixes #13)

tunnckoCore commented on index.js:
```diff
@@ -8,20 +8,21 @@
     name: 'prepack',
-    transform: (source, id) => {
-      if (!filter(id)) return null
+    transform: (fileContents, filePath) => {
```
You repo is using `transformBundle`, isn't that better? i'm totally agree on changing to it, it make sense.

tunnckoCore replied:
also i think that the second argument is options object? Sorry, i'm may be a bit out of that if that was changed in recent Rollup versions.

joseluisq replied:
> You repo is using transformBundle, isn't that better? i'm totally agree on changing to it, it make sense.

No, but I am not sure how `transformBundle` can receive the filePath because prepack need it.
check out https://github.com/rollup/rollup/wiki/Plugins#creating-plugins

joseluisq replied:
> also i think that the second argument is options object? Sorry, i'm may be a bit out of that if that was changed in recent Rollup versions.

The second argument is the `filePath` string basically.

tunnckoCore replied:
Ah yea, really, it's typically called `id` which is the full filepath, yea.

The thing that we can do is to add the `options() {}` to the return, or it was something similar, i'll check here and there.

tunnckoCore replied:
Yea, as in here `https://github.com/rollup/rollup-plugin-node-resolve/blob/master/src/index.js#L72-L74`

`options() {}` method which gets the options. Probably something along these lines

```js
return {
  options(opts) {
    inputFile = opts.input.file
  }
}
```

tunnckoCore replied:
Oh, no, forget. Your approach in the repo make total sense. Change it as there - plugin accepts the dist file as 1st param, and options as 2nd param.

joseluisq replied:
Exactly!
In that case we can use `transformBundle` in the same way as Uglify does it https://github.com/TrySound/rollup-plugin-uglify/blob/master/index.js#L10

joseluisq replied:
note that in this case I have used `transform` to pass not the filename in the plugins array:

```js
{
  plugins: [
    typescript(),
    prepack({ only options here }),
    uglify(),
    // etc
  ]
}
```

joseluisq replied:
or do you prefer other approach ?

tunnckoCore replied:
Yea, but is Prepack need filepath or not? Is it required? If not so, then we can just allow passing options as first param. Otherwise options should be second.

joseluisq replied:
I have verified it is not required, in our case we use `fileContents` only.

joseluisq replied:
Here my tests:

__1. Using transformBundle:__

`transformBundle` is an alias of `transformChunk(source, outputOptions, chunk)`:
> (formerly transformBundle) chunk transformer function called for each Rollup output chunk file. See https://github.com/rollup/rollup/wiki/Plugins

Received input:

```js
'use strict';

['A', 'B', 42].forEach(function (x) {
  var name = '_' + x.toString()[0].toLowerCase();
});
```

Prepack output:
```js
"use strict";
```

__2. Using transform:__

`transform( source, id )`:
> module transformer function. See https://github.com/rollup/rollup/wiki/Plugins

Received input:
```js
var self = {}
;['A', 'B', 42].forEach(function (x) {
  var name = '_' + x.toString()[0].toLowerCase()
  var y = parseInt(x)
  self[name] = y || x
})
```

Prepack output:
```js
(function () {
  var _$0 = this;

  var _0 = {
    _a: "A",
    _b: "B",
    _4: 42
  };
  _$0.self = _0;
}).call(this);
```

Conclusion: In this case we need to use `transform` function because it's applied to the module rather than chunk.

tunnckoCore commented on index.js:
```diff
@@ -8,20 +8,21 @@
 module.exports = function rollupPluginPrepack (options) {
-  options = Object.assign({ include: '**/*.js' }, options)
-
-  const filter = utils.createFilter(options.include, options.exclude)
```
Why removing the filtering?

joseluisq replied:
> Why removing the filtering?

Because these are not supported in the last version, check out the options https://prepack.io/getting-started.html

tunnckoCore replied:
What you mean by not supported. It needs javascript file, so we need to filter. But the above comments defeat this discussions since we will require to pass dist file as in your repo.

joseluisq replied:
Are `include` and `exclude` part of Prepack options?

tunnckoCore replied:
no, they are part of rollup

tunnckoCore commented on index.js:
```diff
@@ -8,20 +8,21 @@
-
-  const filter = utils.createFilter(options.include, options.exclude)
-  const handle = (str) => `export default ${JSON.stringify(str.trim())}`
+  const handle = (str) => `export default ${str.trim()}`
```
Isn't more sure with JSON.stringify? I've seen in it couple of other plugins. And still think that it make sense.

joseluisq replied:
> Isn't more sure with JSON.stringify? I've seen in it couple of other plugins. And still think that it make sense.

Sure, We can use `JSON.stringify`

tunnckoCore reviewed (changes requested):
thoughts on the comments needed

Pyrolistical commented on index.js:
```diff
@@ -8,20 +8,21 @@
 
-      let { code, map } = prepack.prepack(source, options)
+      const { code, map } = prepack.prepackSources(sources, options)
       return { code: handle(code), map }
```
can we just inline `handle`?

tunnckoCore replied:
Yea, but just style preference and some personal habits.

joseluisq replied:
> can we just inline handle?

what do you mean, an example?

joseluisq committed:
improve transform (fixes #13)

- JSON.stringify usage
- prepack fileContents param only
- try/catch staments
- add testing case for errors
- fix testing case for rollup generate test
- removed this in fixure sample according to https://github.com/rollup/rollup/issues/794

joseluisq reviewed (commented):
This commit improves the transform functionality:

- JSON.stringify usage
- prepack` fileContents` param only
- try/catch statements
- fix testing case for rollup generate test
- add testing case for errors
- removed `this` keyword in fixture according to rollup/rollup#794

tunnckoCore commented:
Looks good now. I can merge later today.

joseluisq commented:
Fine!
When this occurs I will deprecate my module in favor of yours.

thisconnect commented:
> later today 

today? :)

tunnckoCore commented:
dooohh, sorry for the delay... i'll try to do it soon.

tunnckoCore commented:
Hell... Why I always forgetting. I merged it for now, publish isn't guaranteed now or in the couple of days.

@joseluisq  thanks anyway :)

tunnckoCore commented:
Published 1.1.0 stable.
